### PR TITLE
adds jinja2 filters for working with network devices

### PR DESCRIPTION
**Pull request by privateip**
* adds network textfsm filters
* adds network config filter

### CODE REVIEW

File: docs/docsite/rst/playbooks_filters.rst [hunk 1]
```diff
@@ -318,6 +318,35 @@ address. For example, to get the IP address itself from a CIDR, you can use::
...
 in :doc:`playbooks_filters_ipaddr`.
 
+.. _network_filter:
+
+Network filters
+```````````````
+
+.. versionadded:: 2.4
+
+To generate a list of command diffs for a network configuration:
```
Comment (gundalow) [inline docs/docsite/rst/playbooks_filters.rst:328]: For code blocks in RST you use `::`, blank line, four spaces

e.g.
```
To generate a list of command diffs for a network configuration::

    {{ myvar | diff_network_config(current_config, indent=1 }}

````

File: lib/ansible/plugins/filter/network.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+from ansible.module_utils.netcfg import NetworkConfig, dumps
+from ansible.errors import AnsibleError
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+NETWORK_OS = {
```
Comment (gundalow) [inline lib/ansible/plugins/filter/network.py:43]: I wonder if this should be called `NETWORK_OS_INDENT`, or similar

File: lib/ansible/plugins/filter/network.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+def parse_cli(value, template):
+    if not HAS_TEXTFSM:
+        raise AnsibleError('parse_cli filter requires TextFSM library to be installed')
+
+    url = urlparse.urlparse(template)
+    if url.scheme.startswith('http'):
+        try:
+            handler = {}
+            if 'HTTP_PROXY' in os.environ:
```
Comment (gundalow) [inline lib/ansible/plugins/filter/network.py:75]: Might be worth adding some comments to say why you set the proxy to localhost.

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by jedelman8** [comment]
Glad to see TextFSM integration going into core.

Few questions for now.

* Are you doing anything with respect a template repository?  I hope not.
* Are you open to having more options in the filter such as returning a list of dictionaries and normalizing the TextFSM object a bit more.
* Are you open to using the `clitable` and add inputs such as template directory, command, and platform?

File: docs/docsite/rst/playbooks_filters.rst [hunk 4]
```diff
@@ -318,6 +318,35 @@ address. For example, to get the IP address itself from a CIDR, you can use::
...
+
+  {{ myvar | diff_network_config(current_config, indent=1 }}
```
Comment (itdependsnetworks) [inline docs/docsite/rst/playbooks_filters.rst:330]: need to close out the function: e.g. 
`{{ myvar | diff_network_config(current_config, indent=1 ) }}`

Also might be a good idea to show example using a file lookup.

File: lib/ansible/plugins/filter/network.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    'eos': 4
+}
+
+def diff_network_config(value, base, network_os=None, indent=1, match='line', replace='line'):
+
+    assert match in ('line', 'strict', 'exact', 'none')
+    assert replace in ('line', 'block')
+    assert network_os in NETWORK_OS.keys()
+
+    if network_os:
```
Comment (itdependsnetworks) [inline lib/ansible/plugins/filter/network.py:56]: Not sure how this would ever hit base on the previous assert? Perhaps the assert should be here?

**Comment by itdependsnetworks** [comment]
Just want to confirm there is no namespace issue with filter_plugins, like there is with modules, e.g. the setup.py issue. Otherwise network.py may be too generic.

**Comment by privateip** [comment]
@jedelman8 

> Are you doing anything with respect a template repository? I hope not.

No, nothing planned

> Are you open to having more options in the filter such as returning a list of dictionaries and normalizing the TextFSM object a bit more.

Can you elaborate more what you are thinking here? 

> Are you open to using the clitable and add inputs such as template directory, command, and platform?

I'm not sure this is needed as we the same thing can be easily accomplished in the playbook  unless I'm not completely understanding what you are asking here.  WRT `clitable` sure

**Comment by privateip** [comment]
@itdependsnetworks 

> Just want to confirm there is no namespace issue with filter_plugins, like there is with modules, e.g. the setup.py issue. Otherwise network.py may be too generic.

filters same as modules aren't namespaced, not sure what the concern here is

**Comment by gundalow** [comment]
I wonder if it would be worth adding this to `CHANGELOG.md` as well

**Pull request closed without merging**